### PR TITLE
[move][vm] Prevent to-module transformation in freeze

### PR DESCRIPTION
**Pull request by modocache**
## Motivation (updated Jun 17 3:43PM)

The Move bounds checker was written to operate upon on a module, so to perform bounds-checking on a script involved transforming it into a module. This script-to-module transformation results in instantiations of `<SELF>` identifiers, dummy addresses, and function handles for the 'main' function (the only function that can exist in a script).
    
However, we can eliminate the need for all this transformation logic by instead rewriting the bounds checker to operate upon a facade like `BinaryIndexedView`. ~~The only snag is that `BinaryIndexedView` wraps compiled scripts and modules that have already been bounds-checked and verified. So, add a new type, `BinaryIndexedViewMut`, that wraps `CompiledScriptMut` and `CompiledModuleMut`.~~
    
The bounds checker operates on collections, and not on objects at specific indices, so ~~`BinaryIndexedViewMut` implements~~ add `BinaryIndexedView` methods to access those collections. For collections that don't exist on scripts, such as struct definitions, the wrapper ~~returns an empty slice~~ returns `None`.

Originally the motivation for this change was to address #8500, but this change alone doesn't prevent the crash seen in that report. I do think this change will make it easier to implement a clean fix for that issue, though.

## Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

I added some tests basic tests that mimic existing bound checker tests in bytecode-verifier-tests.

### CODE REVIEW

**Comment by mengxu-fb** [comment]
cc-ing @darioncassel due to the relation to #8500

File: language/bytecode-verifier/bytecode-verifier-tests/src/unit_tests/bounds_tests.rs [hunk 1]
```diff
@@ -124,6 +129,20 @@ fn invalid_locals_id_in_call() {
...
+    use Bytecode::*;
+
+    let mut s = basic_test_script();
+    s.function_instantiations.push(FunctionInstantiation {
+        handle: FunctionHandleIndex::new(0),
+        type_parameters: SignatureIndex::new(1),
+    });
+    let func_inst_idx = FunctionInstantiationIndex(s.function_instantiations.len() as u16 - 1);
+    s.code.code = vec![CallGeneric(func_inst_idx)];
+    s.freeze().unwrap_err();
```
Comment (sblackshear) [inline language/bytecode-verifier/bytecode-verifier-tests/src/unit_tests/bounds_tests.rs:143]: It might be useful to check that we get the expected error code (here and other tests). It looks like the existing tests don't do this, but shouldn't hurt to be more specific (especially given that the error codes need to be stable).

Comment (modocache) [reply language/bytecode-verifier/bytecode-verifier-tests/src/unit_tests/bounds_tests.rs:143]: Good idea, thanks! See #8537.

File: language/bytecode-verifier/bytecode-verifier-tests/src/unit_tests/bounds_tests.rs [hunk 2]
```diff
@@ -157,6 +192,23 @@ fn invalid_struct_as_type_actual_in_exists() {
     m.freeze().unwrap_err();
 }
 
+#[test]
+fn script_invalid_struct_as_type_actual_in_exists() {
```
Comment (vgao1996) [inline language/bytecode-verifier/bytecode-verifier-tests/src/unit_tests/bounds_tests.rs:196]: FYI: we had a historical debate on terminologies: `(type) formal + (type) actual` vs `(type) parameter + (type) argument`. We didn't reach a conclusion immediately, but over time, at least from my view, the latter became the de facto winner. It would be better if we would keep our usages consistent.

**Review by vgao1996 — APPROVED** [review]
Overall this looks good to me!

I wonder if you should add another test case: 
- A script with `n` module handles, but contains `ModuleHandleIndex(n)`. If I'm understanding it correctly, the original script to module conversion would not handle this correctly, since it may create a dummy module handle for Self.

File: language/bytecode-verifier/bytecode-verifier-tests/src/unit_tests/bounds_tests.rs (hunk 2 above)
Comment (vgao1996) [reply language/bytecode-verifier/bytecode-verifier-tests/src/unit_tests/bounds_tests.rs:196]: Oh sorry for the confusion, I meant renaming `script_invalid_struct_as_type_actual_in_exists` to `script_invalid_struct_as_type_argument_in_exists`. 

Let's keep the script prefix since it's more explicit.

File: language/move-binary-format/src/binary_views.rs [hunk 3]
```diff
@@ -199,35 +198,158 @@ impl<'a> BinaryIndexedView<'a> {
...
+    }
+
+    pub fn function_handles(&self) -> &[FunctionHandle] {
+        match self {
+            BinaryIndexedViewMut::Module(module) => &module.function_handles,
+            BinaryIndexedViewMut::Script(script) => &script.function_handles,
+        }
+    }
+
+    pub fn field_handles(&self) -> &[FieldHandle] {
```
Comment (tnowacki) [inline language/move-binary-format/src/binary_views.rs:285]: I would make this (and several below) an option, but the point is kind of moot...

File: language/move-binary-format/src/binary_views.rs [hunk 4]
```diff
@@ -199,35 +198,158 @@ impl<'a> BinaryIndexedView<'a> {
...
         }
     }
 }
 
+// A `BinaryIndexedViewMut` provides table indexed access for `CompiledModuleMut`
+// and `CompiledScriptMut` -- it is the "mutable" counterpart to `BinaryIndexedView`,
+// with one distinction: this class provides access to the underlying collections
+// within the script or module for bounds checking. Accessing collections that cannot exist on
+// a script results in an empty slice, not an error.
+pub enum BinaryIndexedViewMut<'a> {
```
Comment (tnowacki) [inline language/move-binary-format/src/binary_views.rs:235]: We have an issue/task out to kill the mut versions (maybe... probably) At any rate we really shouldn't be building new frameworks on top of them

**Review by tnowacki — CHANGES_REQUESTED** [review]
Taking an initial look. Just commenting on the main issue and will keep reviewing with suggestions.

File: language/move-binary-format/src/check_bounds.rs [hunk 5]
```diff
@@ -19,100 +20,132 @@ use move_core_types::vm_status::StatusCode;
...
 
 pub struct BoundsChecker<'a> {
-    module: &'a CompiledModuleMut,
+    view: &'a BinaryIndexedViewMut<'a>,
     current_function: Option<FunctionDefinitionIndex>,
 }
 
 impl<'a> BoundsChecker<'a> {
-    pub fn verify(module: &'a CompiledModuleMut) -> PartialVMResult<()> {
+    pub fn verify_script(view: &'a BinaryIndexedViewMut) -> PartialVMResult<()> {
```
[the remark this answers is not in the document]
Comment (modocache) [reply language/move-binary-format/src/check_bounds.rs:28]: Sounds good, I'll try to do this. I went this way because, before speaking with @vgao1996 earlier today, I thought the `CompiledScriptMut` vs `CompiledScript` distinction was a useful one that we wanted to keep in this codebase. Since the only difference between `CompiledScriptMut` and `CompiledScript` is that the latter has been bounds-checked, I thought there simply wasn't a way to get a `CompiledScript` from a `CompiledScriptMut` without calling `CompiledScriptMut::freeze` and performing bounds checking on the `CompiledScriptMut`.

But now, knowing the distinction between mut vs non-mut is meant to be eliminated, I may try to eliminate the `CompiledScriptMut::freeze` method altogether, since it seems like the entire point is to run the bounds checker and return a non-mut object.

**Comment by modocache** [comment]
Thanks for the additional comments! Also FWIW, the test failures are due to the fact that bytecode verification runs on the code in the `main` function as part of `BoundsChecker::check_function_def`. Using `CompiledScriptMut::into_module` to translate a script into a module adds a function definition for `main`, resulting in the checks running over the code in `main`. Assuming this patch doesn't create function definitions for `main`, I'll need to modify the `BoundsChecker::check_code` function to be able to run on a script's code, without providing a function definition or definition index.

File: language/move-binary-format/src/check_bounds.rs [hunk 6]
```diff
@@ -19,100 +20,158 @@ use move_core_types::vm_status::StatusCode;
...
+        for script_handle in bounds_check.view.module_handles() {
+            bounds_check.check_module_handle(script_handle)?
+        }
+        for struct_handle in bounds_check.view.struct_handles() {
+            bounds_check.check_struct_handle(struct_handle)?
+        }
+        for function_handle in bounds_check.view.function_handles() {
+            bounds_check.check_function_handle(function_handle)?
+        }
+        for field_handle in bounds_check.view.field_handles().unwrap_or(&[]) {
```
Comment (tnowacki) [inline language/move-binary-format/src/check_bounds.rs:56]: `.iter().flatten()` might be more idiomatic here

File: language/move-binary-format/src/check_bounds.rs [hunk 7]
```diff
@@ -19,100 +20,158 @@ use move_core_types::vm_status::StatusCode;
...
+        for struct_def in bounds_check.view.struct_defs().unwrap_or(&[]) {
+            bounds_check.check_struct_def(struct_def)?
+        }
+        let view = bounds_check.view;
+        for (function_def_idx, function_def) in
+            view.function_defs().unwrap_or(&[]).iter().enumerate()
+        {
+            bounds_check.check_function_def(function_def_idx, function_def)?
+        }
+        bounds_check.check_self_module_handle()
```
Comment (tnowacki) [inline language/move-binary-format/src/check_bounds.rs:80]: It looks like most of this code is shared between the two. Maybe just have a function for all the shared stuff and then keep the non-overlapping code in `verify_module/script`

File: language/move-binary-format/src/check_bounds.rs [hunk 8]
```diff
@@ -19,100 +20,158 @@ use move_core_types::vm_status::StatusCode;
...
+        check_bounds_impl(&self.view.address_identifiers(), module_handle.address)?;
+        check_bounds_impl(&self.view.identifiers(), module_handle.name)
     }
 
     fn check_self_module_handle(&self) -> PartialVMResult<()> {
-        check_bounds_impl(
-            &self.module.module_handles,
-            self.module.self_module_handle_idx,
-        )
+        if let Some(idx) = self.view.self_handle_idx() {
```
Comment (tnowacki) [inline language/move-binary-format/src/check_bounds.rs:144]: nit: pattern match `match self.view.self_handle_idx() { Some(idx) => ... | None => Ok(()) }`

File: language/move-binary-format/src/check_bounds.rs [hunk 9]
```diff
@@ -123,9 +182,14 @@ impl<'a> BoundsChecker<'a> {
     }
 
     fn check_field_handle(&self, field_handle: &FieldHandle) -> PartialVMResult<()> {
-        check_bounds_impl(&self.module.struct_defs, field_handle.owner)?;
+        check_bounds_impl(&self.view.struct_defs().unwrap_or(&[]), field_handle.owner)?;
```
Comment (tnowacki) [inline language/move-binary-format/src/check_bounds.rs:185]: same as above

File: language/move-binary-format/src/check_bounds.rs [hunk 10]
```diff
@@ -123,9 +182,14 @@ impl<'a> BoundsChecker<'a> {
...
         // field offset must be in bounds, struct def just checked above must exist
-        if let Some(struct_def) = &self.module.struct_defs.get(field_handle.owner.into_index()) {
+        if let Some(struct_def) = &self
+            .view
+            .struct_defs()
+            .unwrap_or(&[])
+            .get(field_handle.owner.into_index())
```
Comment (tnowacki) [inline language/move-binary-format/src/check_bounds.rs:191]: maybe add `struct_def_at` and have it return an option

File: language/move-binary-format/src/check_bounds.rs [hunk 11]
```diff
@@ -146,9 +210,12 @@ impl<'a> BoundsChecker<'a> {
         &self,
         struct_instantiation: &StructDefInstantiation,
     ) -> PartialVMResult<()> {
-        check_bounds_impl(&self.module.struct_defs, struct_instantiation.def)?;
         check_bounds_impl(
-            &self.module.signatures,
+            &self.view.struct_defs().unwrap_or(&[]),
```
Comment (tnowacki) [inline language/move-binary-format/src/check_bounds.rs:214]: maybe making something for `check_bounds_impl_opt` that just gives an out of bounds error on `None` might be a bit more grokkable. Took me a bit to think through this working

File: language/move-binary-format/src/check_bounds.rs [hunk 12]
```diff
@@ -208,9 +278,9 @@ impl<'a> BoundsChecker<'a> {
         function_def: &FunctionDefinition,
     ) -> PartialVMResult<()> {
         self.current_function = Some(FunctionDefinitionIndex(function_def_idx as TableIndex));
-        check_bounds_impl(&self.module.function_handles, function_def.function)?;
+        check_bounds_impl(&self.view.function_handles(), function_def.function)?;
         for ty in &function_def.acquires_global_resources {
-            check_bounds_impl(&self.module.struct_defs, *ty)?;
+            check_bounds_impl(&self.view.struct_defs().unwrap_or(&[]), *ty)?;
```
Comment (tnowacki) [inline language/move-binary-format/src/check_bounds.rs:283]: same as above with `bounds_impl_opt`

**Review by tnowacki — COMMENTED** [review]
Looking better!

Few mostly style things:
- I really hate if-let-else and will comment that on any PR where I see it :P pet peeve really
- I find the patterns of `unwrap_or(&[])` a bit confusing, and would like to see some of that cleaned up if possible

**Comment by tnowacki** [comment]
> I'll need to modify the BoundsChecker::check_code function to be able to run on a script's code, without providing a function definition or definition index.

Providing definition index 0 might be fine for that. I think we do that in a few places but I could be wrong

File: language/move-binary-format/src/check_bounds.rs (hunk 10 above)
Comment (modocache) [reply language/move-binary-format/src/check_bounds.rs:191]: Yeah, I think I considered adding or using `*_at` methods in this file, but I decided against it. Looking at the other `BinaryIndexedView::*_at` methods, they all use index expressions `[]` to access the elements directly, and panic if the index is out-of-bounds. That struck me as pretty different from the `get` methods used in this file, which return `Option`.

My take here would be: if we want to clean up some of these expressions by implementing something like `struct_def_at` or (elsewhere in the file) `struct_handle_at`, etc., I would name them differently, to make a distinction between panicking and non-panicking methods. Maybe something like `get_struct_def` or something? What do you think?

**Comment by modocache** [comment]
> I really hate if-let-else and will comment that on any PR where I see it :P pet peeve really

Oh, this totally makes sense to me and I agree. The spots you commented on were oversights on my part, IIRC I tossed in the `else` and forgot to rewrite it cleaner. So thanks for pointing those out!

File: language/move-binary-format/src/check_bounds.rs (hunk 10 above)
Comment (tnowacki) [reply language/move-binary-format/src/check_bounds.rs:191]: maybe all of this is fine? I don't know... I think in general the unwrap_or is a bit weird to me and might prefer `.struct_defs().then(|defs| defs.field_handle.owner.into_index())`

Comment (tnowacki) [reply language/move-binary-format/src/check_bounds.rs:191]: In general, all of the indexed stuff uses `struct_def_at` because it it was guaranteed to be over a bounds-checked module (which isn't the case anymore)

**Comment by vgao1996** [comment]
> Originally the motivation for this change was to address #8500, but this change alone doesn't prevent the crash seen in that report. I do think this change will make it easier to implement a clean fix for that issue, though.
@modocache Could you elaborate on why this alone doesn't prevent the crash? Are there issues other than the problematic script-to-module conversion process?

**Comment by modocache** [comment]
> Could you elaborate on why this alone doesn't prevent the crash? Are there issues other than the problematic script-to-module conversion process?

Sure thing! I'm still learning how to view the bytecode and read MVIR, and what is or isn't considered valid, so bear with me and this wall of text. But, I wonder if the problem here isn't that the round-trip between script-to-module and back again causes an issue, but rather that the provided bytecode itself is invalid. To illustrate, running `cargo run --bin disassembler -- --script --bytecode` on the given file results in this MVIR:

```
module ffffffffffffffffffffffffffffffff.<SELF> {


public <SELF>()
}
```

I compared this to the disassembler output for the `types/src/test_helpers/empty_script.mv` file in this repository, which is as follows:

```
module ffffffffffffffffffffffffffffffff.<SELF> {


public <SELF>() {
B0:
        0: Ret
}
}
```

It appears that the provided file from #8500 has an empty code unit: no basic blocks, no bytecode, no `Ret`. My naive understanding is that this in and of itself ought to be considered invalid, regardless of any bugs that may exist in script-to-module-to-script conversion.

But, moving on to the actual stack trace from the provided file, I see a stack trace that doesn't mention an empty code unit, but rather an out-of-bounds access into a `SignaturePool`. This is that stack trace, which I get by running `cargo run --bin move -- sandbox run`:

```
thread 'main' panicked at 'index out of bounds: the len is 0 but the index is 0', /home/modocache/src/diem/language/move-binary-format/src/access.rs:232:10
stack backtrace:
  <-- snip -->
   6: move_binary_format::access::ScriptAccess::signature_at
             at ./language/move-binary-format/src/access.rs:232:10
   7: bytecode_verifier::binary_views::BinaryIndexedView::signature_at
             at ./language/bytecode-verifier/src/binary_views.rs:80:50
   8: bytecode_verifier::signature::SignatureChecker::check_signature
             at ./language/bytecode-verifier/src/signature.rs:194:23
   9: bytecode_verifier::signature::SignatureChecker::verify_code
             at ./language/bytecode-verifier/src/signature.rs:124:9
  10: bytecode_verifier::signature::SignatureChecker::verify_script_impl
             at ./language/bytecode-verifier/src/signature.rs:49:9
  11: bytecode_verifier::signature::SignatureChecker::verify_script
             at ./language/bytecode-verifier/src/signature.rs:40:9
  12: bytecode_verifier::verifier::verify_script
             at ./language/bytecode-verifier/src/verifier.rs:50:5
  13: move_vm_runtime::loader::Loader::verify_script
             at ./language/move-vm/runtime/src/loader.rs:536:9
  14: move_vm_runtime::loader::Loader::deserialize_and_verify_script
             at ./language/move-vm/runtime/src/loader.rs:510:15
  15: move_vm_runtime::loader::Loader::load_script
             at ./language/move-vm/runtime/src/loader.rs:468:21
  16: move_vm_runtime::runtime::VMRuntime::execute_script
             at ./language/move-vm/runtime/src/runtime.rs:248:13
  17: move_vm_runtime::session::Session<S>::execute_script
             at ./language/move-vm/runtime/src/session.rs:130:9
  18: move_cli::sandbox::commands::run::run
             at ./language/tools/move-cli/src/sandbox/commands/run.rs:125:17
  <-- snip -->
```

When verifying the script's code unit, the signature verifier here is assuming that the code has a signature. Converting a script to a module does add a dummy signature, and it's removed when the module is converted back into a script. My reading of the description in #8500 was that removing this script-to-module-and-back conversion would eliminate the bug, but I don't think that's the case -- I think the dummy signature is added correctly when the script is transformed into a module, and removed correctly when its transformed back into a script. But I wonder, isn't the problem that the original script doesn't have a signature for its code unit when it's deserialized in the first place?

When I first took on this bug, I commented out the line in `CompiledModuleMut::into_script` that removes the dummy signature that's added by `CompiledScriptMut::into_module` (`inner.signatures.pop().unwrap();`), and confirmed that instead of a panic, we get a verification error: `Execution failed with unexpected error EMPTY_CODE_UNIT`. I think this is the correct result: the bytecode has an empty code unit. But the problem isn't that there's a bug in `into_script` that needs to be fixed, but rather the problem seems to be that a script with an empty code unit proceeds to be verified by routines that expect non-empty `SignaturePool`. I think, earlier in the deserialization and verification process, a check that the code unit is empty (and so no `Signature` for the main function exists) could be added, and the bytecode could be rejected sooner. (Or, scripts need to have an empty signature added to their `SignaturePool` outside of the `CompiledScriptMut::into_module` method, and that dummy signature ought not be removed. But this seems incorrect to me.)

This pull request focuses on refactoring, and on removing a script-to-module-to-script round-trip. But, I think I've discovered that such a refactoring won't prevent this panic. Still, this is just my current read of the situation, let me know if I'm misunderstanding something, @vgao1996!

File: language/move-binary-format/src/check_bounds.rs [hunk 13]
```diff
@@ -2,117 +2,141 @@
...
 impl<'a> BoundsChecker<'a> {
-    pub fn verify(module: &'a CompiledModuleMut) -> PartialVMResult<()> {
+    pub fn verify_script(script: &'a CompiledScript) -> PartialVMResult<()> {
+        let mut bounds_check = Self {
+            view: BinaryIndexedView::Script(&script),
+            context: BoundsCheckingContext::Script,
+        };
+        bounds_check.verify_impl()?;
+        // The bounds checker has already checked each function definition's code, but a script's
+        // code exists outside of any function definition. It gets checked here.
```
Comment (vgao1996) [inline language/move-binary-format/src/check_bounds.rs:40]: I think we should check if `self.parameters` is out of bounds here before accessing the signature. I wonder if it would still panic if we get this fixed?

**Review by vgao1996 — COMMENTED** [review]
@modocache thanks for the detailed write up! I now understand this better. I think there are two separate but similar issues here:
1. `CompiledScriptMut` has a field `parameters`, which is an index that points to an element in the signature pool. This was potentially problematic since `to_module` could introduce a new signature, making `parameters` valid even if it was originally out of bounds. Now in this PR, you got rid of the conversion and made the bounds checker work with a `BinaryIndexView`. However if I'm understandingly correct, this alone is insufficient since `parameters` is still unchecked.
2. The problem I was referring to was the `to_module` conversion's potential inability to handle a script with a module handle index that's out of bounds, but just by 1. This is because a new module handed may be added when we convert a script into a module, making the originally invalid index valid. I believe your PR does address this properly.

File: language/move-binary-format/src/check_bounds.rs (hunk 13 above)
Comment (modocache) [reply language/move-binary-format/src/check_bounds.rs]: Thanks for the suggestion and the explanations! I actually did have some code stashed that checks whether the given parameters index is out of bounds, but I wasn't sure it was correct. I uploaded a patch along the lines of what you suggested (I just need to add a regression test).

**Comment by modocache** [comment]
OK, thanks @vgao1996! I added a commit on top of the refactors that also fixes the panic, along with two regression tests.

**Review by vgao1996 — APPROVED** [review]

**Comment by modocache** [comment]
@tnowacki I see you haven't accepted since requesting changes -- hope you don't mind if I land this now, since it makes stacking some other work of mine a LOT easier. Please do comment post-merge with anything else you'd like to see, though, I can patch things in a separate pull request.

**Review by tnowacki — DISMISSED** [review]
Overall looks good! I think some things might need to be cleaned up for errors, but that can be a separate PR. 

Also still not a big fan of the `unwrap_or(&[])` in a few spots, but its minor ;)

**Comment by modocache** [comment]
> Also still not a big fan of the unwrap_or(&[]) in a few spots, but its minor ;)

Ah, good point. Some of them I had a tricky time replacing, let me try again in a separate PR.
